### DISCUSSION TITLE
Custom Metrics Reporter for Elastic APM

### DISCUSSION THREAD

**Discussion by mahipalgurjar** [moleculerjs/moleculer#1089] [Q&A]
Hi All,

Could you please suggest how we can integrate Moleculer Application with  Elastic APM using custom metrics reporter. Any example in this regards will be useful.

**Comment by intech** [comment]
This is an [old version](https://github.com/intech/moleculer-elastic-apm) based on legacy event tracing, demo on these screenshots:
https://github.com/moleculerjs/moleculer/issues/219#issuecomment-384269814

**Reply by mahipalgurjar** [reply]
Hi intech,

Thanks for the reply 
I see the given screenshots but in the screenshots also you can see the same issues as show unknown route like: Get unknown route and POST unknown. also there is issues when hit multiple molecular services it's show all the services as unknown routes. and it's not handle the tracing properly in ELK stack.
Could you please give anyone idea that how can I fixed it.

Thanks

**Comment by scvetkovins** [comment]
Hi @intech I saw your comment in https://github.com/intech/moleculer-elastic-apm/issues/3#issuecomment-571842158
and that you worked on https://github.com/intech/moleculer-elastic-apm in the past.
I am trying to introduce Elastic APM to the moleculer and my services and I am curios to know if you changed stance about this?


Same hi @icebob I did read https://github.com/moleculerjs/moleculer/issues/219#issuecomment-394621436.
Are you maybe familiar if someone achieve integration with Elastic APM to the moleculer, do you maybe have some general advice now if this is possible or your opinion didn't change?

I would be very grateful for your comments and inputs.

**Reply by intech** [reply]
Hi there! Elastic APM has been supporting distributed tracing for a long time now, and many issues have been resolved since its introduction. However, in my Moleculer projects, I've only used the Jaeger module, and it has been sufficient for our needs. In the future, I highly recommend [switching to OpenTelemetry](https://github.com/moleculerjs/moleculer/discussions/1125), especially [when using Elastic APM](https://www.elastic.co/guide/en/observability/current/apm-open-telemetry.html).

P.S.: You [can configure trace collection](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/elasticsearchexporter) in Elastic APM infrastructure via otel collector

**Reply by scvetkovins** [reply]
Hey @intech first of all thanks for your fast answer.
I took a bit time to digest and I had some delay in my work.

> in my Moleculer projects, I've only used the Jaeger module, and it has been sufficient for our needs

I assume that you are referencing that you are using https://github.com/moleculerjs/moleculer-metrics/blob/master/packages/moleculer-jaeger/README.md or you are directly using it like https://moleculer.services/docs/0.14/tracing.html#Jaeger ?

> In the future, I highly recommend https://github.com/moleculerjs/moleculer/discussions/1125, especially [when using Elastic APM](https://www.elastic.co/guide/en/observability/current/apm-open-telemetry.html).

> P.S.: You [can configure trace collection](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/elasticsearchexporter) in Elastic APM infrastructure via otel collector

Thank you for sharing this with me, I guess those things can be only in use once https://github.com/moleculerjs/moleculer/discussions/1125#discussioncomment-6683290 is production ready? Or do you feel that I should try to use middleware inside my code? I guess it would be similar like that or https://github.com/icebob/moleculer-opentelemetry-demo/tree/master

**Reply by intech** [reply]
> or you are directly using it like https://moleculer.services/docs/0.14/tracing.html#Jaeger ?

I'm using the direct option.

When it comes to the open telemetry demo repository, it's better to ask @icebob.

I'm referring to using Jaeger & Prometheus in Moleculer, and the otel-collector for transforming the output into open telemetry format https://opentelemetry.io/docs/collector/configuration/#receivers

**Reply by icebob** [reply]
@scvetkovins you can use the middleware of the moleculer-opentelemetry-demo, it's the same what's included into the next branch. Moleculer 0.15 is in public beta stage, but OTEL tracing won't be added to it, just as an external middleware.

**Reply by scvetkovins** [reply]
Thank you @intech and @icebob for all inputs 🙌
